python/ray/_private/gcs_utils.py

```diff
@@ -182,13 +196,227 @@ def internal_kv_keys(self, prefix: bytes) -> List[bytes]:
+            channel = create_gcs_aio_channel(address)
+        else:
+            assert channel is not None, \
+                "One of address and channel must be specified"
+        self._stub = gcs_service_pb2_grpc.InternalPubSubGcsServiceStub(channel)
+
+    async def publish_error(self, key_id: bytes,
+                            error_info: ErrorTableData) -> None:
+        """Publishes error info to GCS."""
+        msg = pubsub_pb2.PubMessage(
+            channel_type=pubsub_pb2.RAY_ERROR_INFO_CHANNEL,
+            key_id=key_id,
+            error_info_message=error_info)
+        req = gcs_service_pb2.GcsPublishRequest(pub_messages=[msg])
+        await self._stub.GcsPublish(req)
+
+
+class GcsAioSubscriber:
+    """Subscriber to GCS. Uses async io."""
+
+    def __init__(self, address: str = None, channel: grpc.aio.Channel = None):
+        if address:
+            assert channel is None, \
+                "address and channel cannot both be specified"
+            channel = create_gcs_aio_channel(address)
+        else:
+            assert channel is not None, \
+                "One of address and channel must be specified"
+        self._stub = gcs_service_pb2_grpc.InternalPubSubGcsServiceStub(channel)
+        self._subscriber_id = bytes(
+            bytearray(random.getrandbits(8) for _ in range(28)))
+        # Whether error info has been subscribed.
+        self._subscribed_error = False
+        # Buffer for holding error info.
+        self._errors = deque()
+
+    async def subscribe_error(self) -> None:
+        """Starts a subscription for error info"""
+        cmd = pubsub_pb2.Command(
+            channel_type=pubsub_pb2.RAY_ERROR_INFO_CHANNEL,
+            subscribe_message={})
+        req = gcs_service_pb2.GcsSubscriberCommandBatchRequest(
+            subscriber_id=self._subscriber_id, commands=[cmd])
+        await self._stub.GcsSubscriberCommandBatch(req)
+        self._subscribed_error = True
+
+    async def poll_error(self, timeout=None) -> Tuple[bytes, ErrorTableData]:
+        """Polls for new error messages."""
+        if not self._subscribed_error:
+            self.subscribe_error()
+
+        if len(self._errors) == 0:
+            req = gcs_service_pb2.GcsSubscriberPollRequest(
+                subscriber_id=self._subscriber_id)
+            reply = await self._stub.GcsSubscriberPoll(req, timeout=timeout)
+            for msg in reply.pub_messages:
+                self._errors.append((msg.key_id, msg.error_info_message))
+
+        if len(self._errors) == 0:
+            return None, None
+        return self._errors.popleft()
+
+        req = gcs_service_pb2.GcsSubscriberPollRequest(
+            subscriber_id=self._subscriber_id)
+        reply = await self._stub.GcsSubscriberPoll(req)
+        error_info = []
+        for msg in reply.pub_messages:
+            error_info.append((msg.key_id, msg.error_info_message))
+        return error_info
```

Comment:
Is there any way we can de-duplicate the aio and non aio subscribers code?

Reply:
Created a base class to pull in request generation logic. These logic will become more complicated once we start adding fault tolerant logic. The remaining logic in each class should hopefully not become more complex.